### ISSUE TITLE
Memory leak after upgrading to the v4.21.0

### ISSUE THREAD

**Issue by JeremieDoctrine** [DataDog/dd-trace-js#3887]
**Expected behaviour**
Expecting a flat memory

**Actual behaviour**
![CleanShot 2023-12-15 at 12 02 46@2x](https://github.com/DataDog/dd-trace-js/assets/59561370/e5295b6b-3ee8-47d9-b493-5129f71fe885)

**Steps to reproduce**

**Environment**
Node.js backend. With express and NestJS

* **Operation system:**
* **Node.js version:** 20.9
* **Tracer version:** 4.21.0
* **Agent version:**
* **Relevant library versions:**

**Comment by scandar** [comment]
I'm facing the same problem.

**Comment by ericyhwang** [comment]
Specifically, the `runtime.node.mem.external` metric is showing a leak:
<img width="761" alt="Screenshot 2023-12-15 at 12 23 13 PM" src="https://github.com/DataDog/dd-trace-js/assets/6915076/63cb95e5-97b1-4d5a-8c71-17cd66d754ec">

The first and third deployments are running dd-trace@4.20.0 and show stable `mem.external`, while the middle deployment is running dd-trace@4.21.0, with a significant memory leak. The third deployment is a hotfix, where the only change from the second deployment is pinning to dd-trace@4.20.0.

It's not just a reporting artifact, either - the drops in the screenshot were from Kubernetes automatically restarting containers due to excessive memory usage.

Per the [Node docs](https://nodejs.org/api/process.html#processmemoryusage):
> `external` refers to the memory usage of C++ objects bound to JavaScript objects managed by V8.
> `arrayBuffers` refers to memory allocated for ArrayBuffers and SharedArrayBuffers, including all Node.js [Buffer](https://nodejs.org/api/buffer.html)s. This is also included in the `external` value.

Package changes between the two versions:
* `dd-trace` - 4.20.0 to 4.21.0
* `@datadog/native-appsec` - 4.0.0 to 5.0.0
* `@datadog/pprof` - 4.0.1 to 4.1.0
* (New) `tlhunter-sorted-set@0.1.0`

**Comment by tlhunter** [comment]
Can you try running your app with `DD_TRACE_SPAN_LEAK_DEBUG=1` and letting me know what the output is? It's probably best to run in staging/locally due to performance overhead. The output should tell us which integration might be causing a leak.

**Comment by ericyhwang** [comment]
> Can you try running your app with `DD_TRACE_SPAN_LEAK_DEBUG=1` and letting me know what the output is? It's probably best to run in staging/locally due to performance overhead. The output should tell us which integration might be causing a leak.

Here's a couple representative snippets from a staging server with `DD_TRACE_SPAN_LEAK_DEBUG=1`.

<details>
<summary>Click here to show logs with some active long-poll requests</summary>

Here, it was getting around 10 long-poll requests per minute from clients, where the server holds open each request for 20 seconds and responds with a heartbeat if there's nothing to notify clients about, which was the case. Trimmed JSON lines are response logs for the Express requests.

Its `mem.external` was linearly growing by around 60 MiB over the course of 1.5 hours.

```text
expired spans: web.request: 1 express.middleware: 51 redis.command: 4 sharedb.request: 2
{"@timestamp":"2023-12-19T05:32:26.327Z",...}
expired spans:
expired spans:
expired spans:
expired spans: express.middleware: 51 web.request: 1 redis.command: 4 sharedb.request: 2
{"@timestamp":"2023-12-19T05:32:29.703Z",...}
expired spans:
expired spans:
expired spans:
expired spans:
expired spans:
expired spans:
expired spans:
expired spans:
expired spans:
expired spans:
expired spans:
expired spans:
expired spans: dns.lookup: 3
{"@timestamp":"2023-12-19T05:32:43.399Z",...}
expired spans: web.request: 5 express.middleware: 109 redis.command: 9 sharedb.request: 4 dns.lookup: 3
{"@timestamp":"2023-12-19T05:32:44.517Z",...}
expired spans: dns.lookup: 3
expired spans: dns.lookup: 3
expired spans: web.request: 2 express.middleware: 51 redis.command: 4 sharedb.request: 2 dns.lookup: 3
{"@timestamp":"2023-12-19T05:32:46.741Z",...}
expired spans: dns.lookup: 3
expired spans: express.middleware: 14 web.request: 2 dns.lookup: 3
expired spans: express.middleware: 51 web.request: 2 redis.command: 4 sharedb.request: 2 dns.lookup: 3
{"@timestamp":"2023-12-19T05:32:49.794Z",...}
expired spans: mongodb.query: 6 dns.lookup: 3
expired spans: dns.lookup: 3
expired spans: dns.lookup: 3
expired spans: web.request: 1 express.middleware: 7 dns.lookup: 3
expired spans: dns.lookup: 3
expired spans: dns.lookup: 3
expired spans: dns.lookup: 3
expired spans: dns.lookup: 3
expired spans: express.middleware: 7 web.request: 1 dns.lookup: 3
expired spans: dns.lookup: 3
expired spans: mongodb.query: 6 dns.lookup: 3
expired spans: dns.lookup: 3
expired spans: dns.lookup: 3
{"@timestamp":"2023-12-19T05:33:03.513Z",...}
expired spans: express.middleware: 65 web.request: 4 redis.command: 4 sharedb.request: 2 dns.lookup: 3
{"@timestamp":"2023-12-19T05:33:04.617Z",...}
expired spans: express.middleware: 51 web.request: 2 redis.command: 4 sharedb.request: 2 dns.lookup: 3
expired spans: dns.lookup: 3
expired spans: express.middleware: 51 web.request: 2 redis.command: 4 sharedb.request: 2 dns.lookup: 3
{"@timestamp":"2023-12-19T05:33:07.234Z",...}
expired spans: dns.lookup: 3
expired spans: express.middleware: 7 web.request: 1 dns.lookup: 3
expired spans: dns.lookup: 3
{"@timestamp":"2023-12-19T05:33:10.295Z",...}
expired spans: web.request: 2 express.middleware: 51 redis.command: 4 sharedb.request: 2 mongodb.query: 6 dns.lookup: 3
expired spans: dns.lookup: 3
expired spans: dns.lookup: 3
expired spans: express.middleware: 7 web.request: 1 dns.lookup: 3
expired spans: dns.lookup: 3
expired spans: dns.lookup: 3
expired spans: dns.lookup: 3
expired spans: dns.lookup: 3
expired spans: express.middleware: 14 web.request: 2 dns.lookup: 3
expired spans: dns.lookup: 3
expired spans: mongodb.query: 6 dns.lookup: 3
expired spans: dns.lookup: 3
expired spans: dns.lookup: 3
```
</details>

<details>
<summary>Click here to show logs with no client requests</summary>

Here, it was getting no incoming requests from clients, though it was still getting health check requests from Kubernetes. The server also does some inexpensive background polling of Mongo.

While "idle", the `mem.external` grew by 4 MiB over 15 minutes, so still growing.

```
expired spans: dns.lookup: 3
expired spans: express.middleware: 7 web.request: 1 dns.lookup: 3
expired spans: dns.lookup: 3
expired spans: dns.lookup: 3
expired spans: dns.lookup: 3
expired spans: dns.lookup: 3
expired spans: express.middleware: 7 web.request: 1 dns.lookup: 3
expired spans: dns.lookup: 3
expired spans: dns.lookup: 3
expired spans: mongodb.query: 6 dns.lookup: 3
expired spans: dns.lookup: 3
expired spans: express.middleware: 14 web.request: 2 dns.lookup: 3
expired spans: dns.lookup: 3
expired spans: dns.lookup: 3
expired spans: dns.lookup: 3
expired spans: dns.lookup: 3
expired spans: express.middleware: 7 web.request: 1 dns.lookup: 3
expired spans: dns.lookup: 3
expired spans: dns.lookup: 3
expired spans: mongodb.query: 6 dns.lookup: 3
```
</details>

**Comment by luishrodg** [comment]
I'm facing the same problem.

**Comment by nubunto** [comment]
we've seen the following in an internal application:


```
node:internal/streams/readable:957
Readable.prototype.removeListener = function(ev, fn) {
                                            ^

RangeError: Maximum call stack size exceeded
    at Readable.removeListener (node:internal/streams/readable:957:45)
    at Socket.setStreamTimeout [as setTimeout] (node:internal/stream_base_commons:253:12)
    at socket.setTimeout (/node_modules/.pnpm/dd-trace@4.21.0/node_modules/dd-trace/packages/datadog-instrumentations/src/http/client.js:90:41)
    at socket.setTimeout (/node_modules/.pnpm/dd-trace@4.21.0/node_modules/dd-trace/packages/datadog-instrumentations/src/http/client.js:90:41)
    at socket.setTimeout (/node_modules/.pnpm/dd-trace@4.21.0/node_modules/dd-trace/packages/datadog-instrumentations/src/http/client.js:90:41)
    at socket.setTimeout (/node_modules/.pnpm/dd-trace@4.21.0/node_modules/dd-trace/packages/datadog-instrumentations/src/http/client.js:90:41)
    at socket.setTimeout (/node_modules/.pnpm/dd-trace@4.21.0/node_modules/dd-trace/packages/datadog-instrumentations/src/http/client.js:90:41)
    at socket.setTimeout (/node_modules/.pnpm/dd-trace@4.21.0/node_modules/dd-trace/packages/datadog-instrumentations/src/http/client.js:90:41)
    at socket.setTimeout (/node_modules/.pnpm/dd-trace@4.21.0/node_modules/dd-trace/packages/datadog-instrumentations/src/http/client.js:90:41)
    at socket.setTimeout (/node_modules/.pnpm/dd-trace@4.21.0/node_modules/dd-trace/packages/datadog-instrumentations/src/http/client.js:90:41)

Node.js v18.19.0
```

**Comment by Drejerdk** [comment]
I recently upgraded to v4.21, and since experienced a leak with several thousands ArrayBuffers being retained in memory.
After forcing it back to v4.17.0 I'm not seeing any issues.

**Comment by minimalmonkey** [comment]
Facing the same issue. I got paged at 1am because of this with our app memory reaching a high level.

**Comment by tlhunter** [comment]
@ericyhwang thanks for the output! Do you know if you have incoming requests that exceed 60 seconds?

**Comment by ericyhwang** [comment]
> Do you know if you have incoming requests that exceed 60 seconds?

Shouldn't be any. On the staging server showing the leaks, there were definitely not any exceeding 20 seconds.

**Comment by Qard** [comment]
The `setTimeout` leak has been fixed in [v4.22.0](https://github.com/DataDog/dd-trace-js/releases/tag/v4.22.0) and [v3.43.0](https://github.com/DataDog/dd-trace-js/releases/tag/v3.43.0). This is likely also the source of `ArrayBuffer` instances being held.

I will close this now, but if you find your leak is still not resolved please open a new issue with the details of your specific leak case.
